docs/architecture/adr-003-dynamic-capability-store.md

```diff
@@ -0,0 +1,270 @@
+  memoryStore := ctx.KVStore(ck.memoryKey)
+  // initialise memory store for all names in persistent store
+  for key, value := range persistentStore.Iter() {
+    capability = &CapabilityKey{name: key}
+    memoryStore.Set("fwd/" + capability, name)
+    memoryStore.Set("rev/" + name, capability)
+  }
+  ck.sealed = true
+}
+```
+
+`NewCapability` can be called by any module to create a new unique, unforgeable object-capability
+reference. The newly created capability is *not* automatically persisted, `ClaimCapability` must be
+called on the `ScopedCapabilityKeeper` in order to persist it.
+
+```golang
+func (sck ScopedCapabilityKeeper) NewCapability(ctx Context, name string) (Capability, error) {
+  memoryStore := ctx.KVStore(sck.memoryKey)
+  // check name not taken in memory store
+  if memoryStore.Get("rev/" + name) != nil {
+    return nil, errors.New("name already taken")
+  }
+  // create a new capability
+  capability := &CapabilityKey{name: name}
+  // set forward mapping in memory store from capability to name
+  memoryStore.Set("fwd/" + capability, name)
+  // set reverse mapping in memory store from name to capability
+  memoryStore.Set("rev/" + name, capability)
+  // return the newly created capability
+  return capability
+}
+```
+
+`AuthenticateCapability` can be called by any module to check that a capability
+does in fact correspond to a particular name (the name can be untrusted user input).
+
+```golang
+func (sck ScopedCapabilityKeeper) AuthenticateCapability(name string, capability Capability) bool {
+  memoryStore := ctx.KVStore(sck.memoryKey)
+  // return whether forward mapping in memory store matches name
+  return memoryStore.Get("fwd/" + capability) === name
+}
+```
+
+`ClaimCapability` allows a module to claim a capability key which it has received (perhaps by calling `NewCapability`, or from another module), so that future `GetCapability` calls will succeed.
+
+`ClaimCapability` MUST be called, even if `NewCapability` was called by the same module. Capabilities are single-owner, so if multiple modules have a single `Capability` reference, the last module
+to call `ClaimCapability` will own it. To avoid confusion, a module which calls `NewCapability` SHOULD either call `ClaimCapability` or pass the capability to another module which will then claim it.
+
+```golang
+func (sck ScopedCapabilityKeeper) ClaimCapability(ctx Context, capability Capability) error {
+  persistentStore := ctx.KVStore(sck.persistentKey)
+  memoryStore := ctx.KVStore(sck.memoryKey)
+  // fetch name from memory store
+  name := memoryStore.Get("fwd/" + capability)
+  if name === nil {
+    return errors.New("capability not found")
+  }
+  // set name to module in persistent store
+  persistentStore.Set(name, sck.moduleName)
+}
+```
+
+`GetCapability` allows a module to fetch a capability which it has previously claimed by name. The module is not allowed to retrieve capabilities which it does not own.
```

Comment:
If another module claims the capability by name, then this module will no longer be able to retrieve capability correct?

Reply:
Answered further down, but think it might be more useful having answer here

Reply:
Aye. Updated.